citesphere/src/main/java/edu/asu/diging/citesphere/web/HomeController.java

```diff
@@ -26,14 +32,20 @@ public String home(Authentication authentication, Model model) {
             boolean isConnected = tokenManager.getToken(user) != null;
             model.addAttribute("isZoteroConnected", isConnected);
             model.addAttribute("user", user);
-            if (isConnected) {
-                model.addAttribute("groups", citationManager.getGroups(user));
-            }
         }
         
         return "home";
     }
     
+    @RequestMapping(value = "/groups")
+    public @ResponseBody List<ICitationGroup> getGroups(Authentication authentication, @RequestParam(defaultValue = "false", required = true, value = "isZoteroConnected") String isConnected) {
+        boolean isZoteroConnected = Boolean.parseBoolean(isConnected);
+        if (isZoteroConnected) {
+            return citationManager.getGroups((IUser) authentication.getPrincipal());
+        }
```

Comment:
this will still time out if it takes too long. there needs to be an async call and the client needs to poll for the results.